File: python/src/problem/leetcode/easy/leetcode_617.py
# coding=utf-8
from src.data_structure.data_structure import TreeNode


class Solution:
    """
    合并二叉树
    """

    def merge_trees(self, t1: TreeNode, t2: TreeNode) -> TreeNode:
        """
        Time: O(n), Space: O(n)
        :param t1:
        :param t2:
        :return:
        """

        if not t1 and not t2:
            return None

        if not t1 or not t2:
            return t1 if t1 else t2

        t1.val = t1.val + t2.val
        t1.left = self.merge_trees(t1.left, t2.left)
        t1.right = self.merge_trees(t1.right, t2.right)
        return t1

File: java/src/main/java/com/ywh/problem/leetcode/easy/LeetCode617.java
package com.ywh.problem.leetcode.easy;

import com.ywh.ds.tree.TreeNode;

/**
 * 合并二叉树
 * [树]
 * 
 * 给定两个二叉树，想象当你将它们中的一个覆盖到另一个上时，两个二叉树的一些节点便会重叠。
 * 你需要将他们合并为一个新的二叉树。合并的规则是如果两个节点重叠，那么将他们的值相加作为节点合并后的新值，
 * 否则不为 NULL 的节点将直接作为新二叉树的节点。
 * 示例 1:
 *      输入:
 *      	Tree 1                     Tree 2
 *                1                         2
 *               / \                       / \
 *              3   2                     1   3
 *             /                           \   \
 *            5                             4   7
 *      输出:
 *      合并后的树:
 *      	     3
 *      	    / \
 *      	   4   5
 *      	  / \   \
 *      	 5   4   7
 *      注意: 合并必须从两个树的根节点开始。
 * 
 * @author ywh
 * @since 2/20/2019
 */
public class LeetCode617 {

    /**
     * Time: O(n), Space: O(n)
     *
     * @param root1
     * @param root2
     * @return
     */
    public TreeNode mergeTrees(TreeNode root1, TreeNode root2) {
        if (root1 == null && root2 == null) {
            return null;
        }
        if (root1 == null || root2 == null) {
            return root1 == null? root2: root1;
        }
        root1.val = root1.val + root2.val;
        root1.left = mergeTrees(root1.left, root2.left);
        root1.right = mergeTrees(root1.right, root2.right);
        return root1;
    }

}
